libservice/src/IpAddressNetlinkChecker.cpp

```diff
@@ -131,11 +139,73 @@ static std::optional<IPv4int> getMappedIPv4Addr(const in6_addr& addr) {
 	return ipv4Binary;
 }
 
+IpAddressNetlinkChecker::ipv6Range IpAddressNetlinkChecker::parseIpv6Range(const std::string& range) const {
+	ipv6Range rangeStruct;
+	const auto slashPos{range.find('/')};
+	rangeStruct.ipv6Address = range.substr(0, slashPos);
+	rangeStruct.prefixLength = slashPos != std::string::npos ? std::stoi(range.substr(slashPos + 1)) : 0;
+	return rangeStruct;
+}
+
+bool IpAddressNetlinkChecker::isInRange(const in6_addr& addr, const std::string& range) const {
+	auto rangeStruct{IpAddressNetlinkChecker::parseIpv6Range(range)};
+
+	in6_addr rangeIpv6Address{};
+	inet_pton(AF_INET6, rangeStruct.ipv6Address.c_str(), &rangeIpv6Address);
+
+	// Create mask
+	in6_addr mask{};
+	for (size_t i = 0; i < sizeof(mask.s6_addr); i++) {
+		if (rangeStruct.prefixLength >= 8) {
+			mask.s6_addr[i] = 0xFF;
+			rangeStruct.prefixLength -= 8;
+		} else if (rangeStruct.prefixLength > 0) {
+			mask.s6_addr[i] = (uint8_t)(0xFF << (8 - rangeStruct.prefixLength));
+			rangeStruct.prefixLength = 0;
+		} else {
+			mask.s6_addr[i] = 0;
+		}
+	}
+
+	// Check mask and addr
+	in6_addr andResult{};
+	for (size_t i = 0; i < sizeof(andResult.s6_addr); ++i) {
+		andResult.s6_addr[i] = addr.s6_addr[i] & mask.s6_addr[i];
+	}
+
+	// Compare andResult with IPv6 address of the rangeStruct
+	return memcmp(&andResult, &rangeIpv6Address, sizeof(andResult)) == 0;
+}
+
+bool IpAddressNetlinkChecker::checkSubnet(
+		const in6_addr& addrToCheck, const in6_addr& interfaceIpv6Addr, const in6_addr& interfaceMask) const {
+	const auto s6Addr32ArraySize = sizeof(addrToCheck.s6_addr32) / sizeof(addrToCheck.s6_addr32[0]);
+	for (size_t i = 0; i < s6Addr32ArraySize; ++i) {
+		if ((addrToCheck.s6_addr32[i] & interfaceMask.s6_addr32[i]) != (interfaceIpv6Addr.s6_addr32[i] & interfaceMask.s6_addr32[i])) {
+			return false;
+		}
+	}
+	return true;
+}
+
 bool IpAddressNetlinkChecker::isV6AddressExternal(const in6_addr& addr) const {
 	if (auto mappedV4Addr = getMappedIPv4Addr(addr); mappedV4Addr) {
 		return isV4AddressExternal(*mappedV4Addr);
 	}
-	throw std::runtime_error("IPv6 is only supported for IPv4 mapped addresses");
+
+	for (auto& ipv6Interface : netlink.collectIpv6Interfaces()) {
```

Comment:
Do I see  correctly? You are calling collectIpv6Interfaces() for each checked addr? In my opinion it should be called only one at the beginning.

Reply:
range expression is always called only once, see https://en.cppreference.com/w/cpp/language/range-for

Reply:
Of course I meant every call _isV6AddressExternal_

Reply:
So you think we should collect IPv6 networks only once? What if they change while the `eBPF-Discovery` program is running?

Reply:
collectIpv6Interfaces can be considered as "heavy"  system call. I think it should be called only once during initialization/startup or  output should be cached.

Reply:
perhaps we should consider running with some long interval, but that is for another story